### ISSUE TITLE
Ability to resume NerDLApproach from checkpoints and saving best model version

### ISSUE THREAD

**Issue by katelynrwalker** [JohnSnowLabs/spark-nlp#2471]
First off, thanks for creating this great library. One potential improvement is that I'd like to be able to train an NERDL model for additional epochs without starting over from the beginning. I read the response to this issue: [1028](https://github.com/JohnSnowLabs/spark-nlp/issues/1028) and the logic for not providing a method to resume training doesn't make sense to me. Perhaps I'm misunderstanding, but that response seems to say that there is no need to fine-tune ever ("we haven't found any practical evidence that someone can resume something over, for instance, CoNLL2003 models in NER and either learn new things or improve it."). Yet this library provides a method for fine-tuning pretrained models via the fit method, right? When I "train" an NERDL model aren't I really doing finetuning of a pretrained model?

I would like to be able to train (fine-tune) a NERDL model for 10 epochs, look at the results, decide I should have trained longer, and then be able to resume training of that same model for an additional 10 epochs, rather than having to start over with a new training for 20 epochs. The current project I'm working on takes several hours to train 10 epochs, so this would be a big time and compute savings!

Or if this is currently possible and I'm just missing something, please let me know.

Thanks!

**Comment by maziyarpanahi** [comment]
Hi @katelynrwalker,

```
When I "train" an NERDL model aren't I really doing finetuning of a pretrained model?
```
Not at all. The .fit() starts the training/learning in all the trainable annotators we have from weights that are all set with zeros (there is no knowledge before to fine-tune it). This continues while providing metrics until the training process is finished and the trained weights will be saved (or used from the memory). It's very important to get the terminology right, you are never training a pre-trained model, you either train a new model from scratch or just download/load a pre-trained model (already trained and saved) and use it just for prediction.

The feature request you mentioned (the feature request) was to download a pre-trained model that was trained on a specific dataset by using a specific embedding so he can fine-tune it on a smaller dataset that could have very different labels, domains, etc. This is entirely different than just simply pausing and just resuming which may improve or may overfit the model (it's always up to users to experiment to get the hyperparameters correct). Although fine-tuning is very popular in many Transformers such as BERT, there is no evidence that it works in NER models. (The NER models are very small even if you trained them on millions of data the saved NER model is 12-14MB compare to 1GB BERT. The fine-tuning can be useful if the model has a very large number of parameters and contains a massive amount of weights which the NER is not like that)

That's being said, what you are asking should be totally possible. However, may not be practical if it comes alone. As I mentioned, if you resume the training from the already saved weights (checkpoints) you may overfit your model and then you will end up with a less accurate model. I would suggest the following features which together would be more helpful:

- Allow saving model after each Epoch on disk (if it's enabled it will save a copy of the trained model after each Epoch)
- Allow resuming training from the last or selected checkpoint (obviously the parameters must be the same, including the number of labels, embeddings dimensions, etc. or the resume will fail)
- Allow saving the best model after each Epoch (there can be another directory that keeps being updated with the best-performed model based on `val_loss` or `f1` for instance like Keras. This way, if you resume and you start seeing a bad performance you will always know where to access the best trained model)

What do you think?

**Comment by katelynrwalker** [comment]
Thank you for the clarification! I was a little fuzzy on what the starting point was for the NERDL annotator - but that makes sense that it starts from zeroes. Is there anywhere I can read more about the architecture used for the NERDL models in spark-nlp? I guess I also misunderstood the other question - my head was too much in my own problem-space!

I agree your suite of features seems more helpful than just resuming training alone. You're right that continued training could cause overfitting, but I suppose overfitting could also happen during an initial training run if maxEpochs is set too high. So it would be really useful to be able to save interim versions of the trained model. (Side question - is there currently an early-stopping feature in NERDL training? I see that you can set a validation proportion or an external test dataset, and min/maxEpochs implies that maybe the training might stop before the max is reached, but if so what statistic is that stopping based on, and is that configurable at all?)

**Comment by maziyarpanahi** [comment]
There is no early stopping feature, the validation split will simply take a chosen percentage from the shuffled training dataset and use that for evaluation metrics during each Epoch. It's for those who don't have dedicated test/dev datasets which is very helpful to monitor the model's performance.

**Comment by Dekermanjian** [comment]
Hello, I am wondering what parameter in NerDLApproach allows saving the model to disk after each epoch? I am looking at the documentation here: https://nlp.johnsnowlabs.com/api/com/johnsnowlabs/nlp/annotators/ner/dl/NerDLApproach 
But I am unsure to which parameter it is. Any help would be appreciated. Thank you.

**Comment by maziyarpanahi** [comment]
This feature doesn’t exist, this is a feature request that hasn’t been implemented yet. 
You can only save the model once the training is over.
